indexing-service/src/main/java/io/druid/indexing/overlord/IndexerMetadataStorageAdapter.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.indexing.overlord;
+
+import com.google.common.base.Preconditions;
+import com.google.inject.Inject;
+import io.druid.java.util.common.DateTimes;
+import org.joda.time.DateTime;
+import org.joda.time.Interval;
+
+import java.util.Comparator;
+import java.util.Optional;
+
+public class IndexerMetadataStorageAdapter
+{
+  private final TaskStorageQueryAdapter taskStorageQueryAdapter;
+  private final IndexerMetadataStorageCoordinator indexerMetadataStorageCoordinator;
+
+  @Inject
+  public IndexerMetadataStorageAdapter(
+      TaskStorageQueryAdapter taskStorageQueryAdapter,
+      IndexerMetadataStorageCoordinator indexerMetadataStorageCoordinator
+  )
+  {
+    this.taskStorageQueryAdapter = taskStorageQueryAdapter;
+    this.indexerMetadataStorageCoordinator = indexerMetadataStorageCoordinator;
+  }
+
+  public int deletePendingSegments(String dataSource, Interval deleteInterval)
+  {
+    // Check the given interval overlaps the interval(minCreatedDateOfActiveTasks, MAX)
+    final Optional<DateTime> minCreatedDateOfActiveTasks = taskStorageQueryAdapter
+        .getActiveTasks()
+        .stream()
+        .map(task -> taskStorageQueryAdapter.getCreatedTime(task.getId()))
+        .min(Comparator.naturalOrder());
+
+    final Interval activeTaskInterval = new Interval(
+        minCreatedDateOfActiveTasks.orElse(DateTimes.MAX),
+        DateTimes.MAX
+    );
+
+    Preconditions.checkArgument(
+        !deleteInterval.overlaps(activeTaskInterval),
+        "Cannot delete pendingSegments because there is at least one running task created at %s",
```

Comment:
hm, I wonder if throwing an exception here is too aggressive, if I understand correctly:
- Since tasks can take a varying amount of time to complete, it seems like it wouldn't be too uncommon to have a situation where there are active tasks, created before the most recently completed task, that are still running
- Would it be better to just return 0 here without an exception, or maybe use `minCreatedDateOfActiveTasks` as the end of the pending segment deletion interval?

Reply:
Good point. I changed to gather all incomplete tasks and the last complete task and find the earliest createdTime. This makes less exceptions occur when the coordinator kills pending segments.

I think throwing an exception here is fine now because it rarely occurs 1) if there is a bug in ```DruidCoordinatorCleanupPendingSegments``` and 2) humans call the overlord API directly with a wrong interval.